### ISSUE TITLE
Configure FAS with openNDS and Node.js server on Raspberry Pi

### ISSUE THREAD

**Issue by theofred** [openNDS/openNDS#304]
This is just a prototype setup to verify a certain use case for my client. For a production build I would probably move over to a router with OpenWrt. However it would still serve a FAS of the same kind. I want to build a Web-App-like next.js Captive Portal that runs locally on a node.js server on the same device as openNDS (for this prototype).

## setup

I've compiled openNDS on my Pi 3 Model B. eth0 is connected to my router. wlan0 is used as AP. Everything is setup so that the default `login_option_enabled` scripts are running as expected. I already could use those CPs with my phone successfully.

The next.js project is just a simple server side rendered page that came straight from the [next.js tutorials page](https://nextjs.org/learn/basics/create-nextjs-app). The server runs on `http://localhost:3000`. When I open Chrome on Pi via remote desktop and browse to this url, I'll get the page.

`opennds.conf`
```sh
# login_option_enabled 0
fasport 3000
faspath /posts/ssg-ssr
fas_secure_enabled 0
faskey 1234567890
fasremoteip 0.0.0.0
fasremotefqdn localhost

debuglevel 3
```

## problem

With an Android  phone I get redirected to the CP-Browser and it opens
```
http://localhost:3000/posts/ssg-ssr?authaction=http://10.3.141.1:2050/opennds_auth/?clientip=10.3.141.228&gatewayname=openNDS&tok=1df1d2
```
but with the error
```
net::ERR_CONNECTION_REFUSED
```

## my guess

Because the client gets redirected to a local IP it gets blocked by the Captive Portal browser. Is that right?
If so, I found a note in other issues stating that `dnsmasq.conf` should be changed to hold a line like
```
address=/#/123.123.123.123
```
However, I don't know what the right steps are to reconfigure opennds so that the CP-browser sees this 'external' IP address and still gets redirected to the local running node.js server.

Is my guess right?

**Comment by theofred** [comment]
@bluewavenet 
Thanks for the quick reply.

`fas_secure_enabled 0` was just for testing, but good to know that it is deprecated.

> I would also suggest you use OpenWrt instead of Raspios for router applications, even if this is a proof of concept (particularly as you intend to move on to proper router hardware). The Pi3 range is fully supported by OpenWrt, as is node.js.
> The Pi's built in wireless chip is designed for STA use only. It will work in AP mode but will be slow and limited to 2 or 3 users simultaneously before dropping packets and getting hot, but I think you are aware of that problem.

Ok, thanks for this limit estimation. That makes the transition definitely more urgent.
Do you know of a good way to enhance the Pi abilities regarding AP without replacing the device  with a router? My client is remote and has a Pi too to test the PoC. Enhancing the Pi AP may be more convenient than trying to integrate a new router into my clients network. And after the PoC the client will decide whether this project goes into production or not, so I'm not sure whether buying a router will be worth it or not.

> In your setup you should have something like: ...

Thank you very much! Now it makes more sense to me and it is working.

I got a second setup where my node.js server is hosted on the internet. Sadly, It returns `ERR_CONNECTION_REFUSED` on that setup. It is hosted on `vercel.app` which hosts many server deployments. My deployment got a unique subdomain but the server IPv4 address is shared between many deployments of course. 
Could that lead to a problem there?

That setup is more like this:
```
fasport 443
faspath /posts/ssg-ssr
fas_secure_enabled 3
faskey 1234567890
fasremoteip 76.76.21.61
fasremotefqdn mydomain.vercel.app
```
With `sudo journalctl -e` I see that `authmon - remote FAS response` return the server page correctly, so openNDS seems to get it. What could be the problem here and how to debug this error? It seems difficult to debug the CP-browser anyways.

**Comment by bluewavenet** [comment]
@theofred 
The FQDN `mydomain.vercel.app` is a CDN, ie it shares not just multiple web sites on one IP, but **_multiple web sites on multiple IP addresses_**.

See:
https://opennds.readthedocs.io/en/stable/fas.html#using-a-cdn-content-delivery-network-hosted-server-for-a-remote-fas

You must set up a **walled garden** for this.
See:
https://opennds.readthedocs.io/en/stable/walledgarden.html#autonomous-walled-garden

**Comment by theofred** [comment]
@bluewavenet 
Thank you for the clarification! I will take a look at it.

**Comment by bluewavenet** [comment]
@theofred 
Feel free to reopen if you have continuing problems with FAS configuration.
If you have problems with CDN or Walled Garden config please open a new issue. ;-)
